### PR TITLE
[Feature] Indexing Discrete and OneHot specs

### PR DESCRIPTION
**Pull request by remidomingues**
## Description

Adding indexing support to `DiscreteTensorSpec` and `OneHotDiscreteTensorSpec` specs. 
- [x] Integers
- [x] Lists
- [x] Tuples
- [x] Ranges
- [x] Slices
- [x] None
- [x] Ellipsis
- [x] Numpy arrays
- [x] Tuples

## Motivation and Context

Addressing feature request of adding indexing to specs: https://github.com/pytorch/rl/issues/1051.

Note: although the feature request states that the desired indexing should behave similarly to tensordict, we have found multiple edge cases where the behavior of Numpy and tensordict differ. For all such cases, the behavior of numpy makes the most sense, it may be that these edge cases are tensordict bugs: \[[1](https://github.com/remidomingues/rl/blob/indexing_specs/test/test_specs.py#L2415), [2](https://github.com/remidomingues/rl/blob/indexing_specs/test/test_specs.py#L2422), [3](https://github.com/remidomingues/rl/blob/indexing_specs/test/test_specs.py#L2430), [4](https://github.com/remidomingues/rl/blob/indexing_specs/test/test_specs.py#L2438), [5](https://github.com/remidomingues/rl/blob/indexing_specs/test/test_specs.py#L2449)\].

## Limitations
- The current pull request only adds indexing to `DiscreteTensorSpec` and `OneHotDiscreteTensorSpec`.
We plan to extend the indexing feature in upcoming PRs but thought it might be best to get it approved in smaller chunks.
- For a [very specific edge case](https://github.com/remidomingues/rl/blob/indexing_specs/test/test_specs.py#L2457), containing a combination of multiple slices, lists and integer indices in a very specific order, our current code matches the behavior of Tensordict, but not of numpy. We'd like to fix that but weren't able to understand the rationale behind such indexing.

## Types of changes

- [x] New feature (non-breaking change which adds core functionality)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: test/test_specs.py [hunk 1]
```diff
@@ -2348,6 +2348,117 @@ def test_to_numpy(self):
...
+    expected_shape = [0] if isinstance(spec, OneHotDiscreteTensorSpec) else []
+    assert empty_spec[...].shape == torch.Size(expected_shape)
+    expected_shape = [2, 4] if isinstance(spec, OneHotDiscreteTensorSpec) else [3, 2]
+    assert spec[..., :2].shape == torch.Size(expected_shape)
+    expected_shape = (
+        [2, 1, 1, 4] if isinstance(spec, OneHotDiscreteTensorSpec) else [3, 2, 1, 1]
+    )
+    assert spec[..., :2, None, None].shape == torch.Size(
+        expected_shape
+    )  # Matches Numpy's behavior. Tensordict would raise "Not enough dimensions".
```
Comment (vmoens) [inline test/test_specs.py:2417]: Thanks for commenting that, but since we want to fix it in tensordict I would rather move the comments in the PR to make them more ephemeral it that makes sense.

File: torchrl/data/tensor_specs.py [hunk 2]
```diff
@@ -30,17 +32,29 @@
...
 
+SHAPE_INDEX_TYPING = Union[
+    int,
+    range,
+    List[int],
+    np.ndarray,
+    Tuple[int, None, slice, List[int], Tuple[Any], type(...)],
+    slice,
+    None,
+    torch.Tensor,
```
Comment (vmoens) [inline torchrl/data/tensor_specs.py:51]: Did we test indexing with a tensor?

**Review by vmoens — APPROVED** [review]
Excellent work, very comprehensive!
Have a look at the two minor comments

File: test/test_specs.py (hunk 1 above)
Comment (remidomingues) [reply test/test_specs.py:2417]: Good call!

Writing for the record that a few tests exhibit a different behavior for numpy and tensordict, especially for indices containing None. We have filed the following [Tensordict issue](https://github.com/pytorch-labs/tensordict/issues/355) and sticked to numpy behavior in these cases.

File: torchrl/data/tensor_specs.py (hunk 2 above)
Comment (remidomingues) [reply torchrl/data/tensor_specs.py:51]: We have a short test here: https://github.com/pytorch/rl/pull/1081/files#diff-9b52a375cf71a2d2d441107d14b011ba181b875085cb5724a19c7f7ac9d56568R2436.

File: test/test_specs.py [hunk 3]
```diff
@@ -2348,6 +2348,112 @@ def test_to_numpy(self):
...
+    assert spec_3d[None, 1, ..., None].shape == torch.Size(expected_shape)
+    # Numpy arrays
+    assert spec[np.array([[1, 2]])].shape == torch.Size([1, 2, 4])
+    # Tensors
+    assert spec[torch.randint(3, (3, 2))].shape == torch.Size([3, 2, 4])
+    # Tuples
+    # Note: nested tuples are supported by specs but transformed into lists, similarity to numpy
+    assert spec_3d[(0, 1), (0, 1)].shape == torch.Size([2, 4])
+    assert spec_3d[:2, (0, 1)].shape == torch.Size([2, 2, 4])
+    assert spec_3d[:2, [0, 1]].shape == torch.Size([2, 2, 4])
```
Comment (vmoens) [inline test/test_specs.py:2436]: to test with a tensor:
```suggestion
    assert spec_3d[:2, (0, 1)].shape == torch.Size([2, 2, 4])
    assert spec_3d[:2, [0, 1]].shape == torch.Size([2, 2, 4])
    assert spec_3d[:2, torch.tensor([0, 1])].shape == torch.Size([2, 2, 4])
```

Comment (remidomingues) [reply test/test_specs.py:2436]: Good call. Support was added for indexing of range, ndarrays and tensors within tuples.

**Review by vmoens — APPROVED** [review]
LGTM
